### PR TITLE
[VAX-532] Introduce dependency info for row add-wins semantics

### PR DESCRIPTION
**Pull request by define-null**
Tags mechanism is added to Satellite protocol to track row dependencies in the form similar to ORSet, to provide add-wins semantics for concurrent operations. Tags stored with the row determine wherever or not row exists.

### CODE REVIEW

File: src/satellite/process.ts [hunk 1]
```diff
@@ -489,26 +717,39 @@ export class SatelliteProcess implements Satellite {
...
   }
 
   async _applyTransaction(transaction: Transaction) {
-    const opLogEntries = fromTransaction(transaction, this.relations)
-
-    await this._apply(opLogEntries, transaction.lsn)
-    this._notifyChanges(opLogEntries)
+    const origin = transaction.origin!
+    if (origin == this._authState!.clientId) {
+      await this._garbageCollectOplog(transaction)
```
Comment (icehaunter) [inline src/satellite/process.ts:737]: Can you please elaborate a little on why this is valid? I'm not sure I follow logic of running a bulk delete operation when applying a transaction, and also limiting the delete operation to the transactions in the same microsecond.

File: src/satellite/process.ts [hunk 2]
```diff
@@ -301,38 +320,67 @@ export class SatelliteProcess implements Satellite {
...
+    if (oplogEntries.length !== 0) {
+      // FIXME: We should probably notify about changes later
+      promises.push(this._notifyChanges(oplogEntries))
 
-    const rows = await this.adapter.query({
-      sql: selectChanges,
-      args: [timestamp],
-    })
-    const results = rows as unknown as OplogEntry[]
+      const shadowEntries = new Map<string, ShadowEntry>()
```
Comment (icehaunter) [inline src/satellite/process.ts:335]: So as far as I can tell, `shadowEntries` is cached only in the scope of this function call, i.e. when the same row is touched multiple times in the same oplog "span", right?

File: src/satellite/process.ts [hunk 3]
```diff
@@ -454,13 +569,126 @@ export class SatelliteProcess implements Satellite {
...
+      args: [timestamp.toISOString(), since],
+    })
+    return rows as unknown as OplogEntry[]
+  }
+
+  async _getOplogShadowEntry(oplog: OplogEntry): Promise<ShadowEntry[]> {
+    const shadow = this.opts.shadowTable.toString()
+    let query
+    let selectTags = `SELECT * FROM ${shadow}`
+    if (oplog != undefined) {
```
Comment (icehaunter) [inline src/satellite/process.ts:599]: If `oplog` can be undefined, let's reflect that in the types of the argument (here and up the call chain). Otherwise, let's drop the condition

**Review by icehaunter — CHANGES_REQUESTED** [review]
Looks cool! I'm still wrapping my head around some shadow entry logic. Meanwhile I had some minor comments and questions

File: src/satellite/client.ts [hunk 4]
```diff
@@ -632,18 +635,20 @@ export class SatelliteClient extends EventEmitter implements Client {
           commit_timestamp: op.begin.commitTimestamp,
           lsn: op.begin.lsn,
           changes: [],
+          origin: op.begin.origin!,
```
[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/client.ts:638]: Good question. So on a direction from Satellite, origin is the same as an id which is passed in the Auth message. So it might very well be ignored by the server. However on the way back it is important to specify that, so that Satellite knows, when it receives transaction which originated from it.

File: src/satellite/process.ts [hunk 5]
```diff
@@ -394,16 +482,32 @@ export class SatelliteProcess implements Satellite {
...
+      incoming
+    )
+
+    /*
+    console.log(local)
+    console.log(incoming)
+
+    console.log('merged:')
+    console.log(merged)
+     */
```
[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/process.ts:510]: If you don't mind I'm gonna clean it up once we are happy with the review. I left it in case we will need more unit tests. I will wait till I get approval from @balegas on that one.

File: src/satellite/process.ts [hunk 6]
```diff
@@ -301,38 +320,67 @@ export class SatelliteProcess implements Satellite {
...
-    const results = rows as unknown as OplogEntry[]
+      const shadowEntries = new Map<string, ShadowEntry>()
+
+      for (const oplogEntry of oplogEntries) {
+        const [cached, shadowEntry] = await this._lookupCachedShadowEntry(
+          oplogEntry,
+          shadowEntries
+        )
+
+        /*
```
[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/process.ts:343]: same as above

File: src/satellite/process.ts (hunk 1 above)
Comment (define-null) [reply src/satellite/process.ts:737]: I should probably add a comment for that in the code but current logic is - to clean oplog when we receive transaction from Electric that corresponds to this part of oplog. Previously we would do such a GC when receiving ack.

File: src/satellite/process.ts (hunk 2 above)
Comment (define-null) [reply src/satellite/process.ts:335]: Yes, that's the idea.

File: src/satellite/process.ts (hunk 1 above)
Comment (icehaunter) [reply src/satellite/process.ts:737]: I'd really appreciate a comment in code about it with this, and probably an explanation as to why is exact equality of items in the oplog on a millisecond time precision is a good indicator of garbage, for posterity

File: src/satellite/merge.ts [hunk 7]
```diff
@@ -33,10 +40,37 @@ export const mergeChangesLastWriteWins = (
...
+}
+
+export const mergeOpTags = (
+  local: OplogEntryChanges,
+  remote: OplogEntryChanges
+): Tag[] => {
+  return calculateTag(local.tag, remote.clear_tags, local.clear_tags)
+}
+
+const calculateTag = (tag: Tag | null, tags: Tag[], clear: Tag[]) => {
```
[the remark this answers is not in the document]
Comment (balegas) [reply src/satellite/merge.ts:61]: We discussed this and I remember agreeing with code, but need to refresh my memory and have things explained to me from your perspective.

When applying incoming operations, we want to add to the local shadow table new incoming tags and make sure that we don't re-add tags that have been cleared by local operations (this is to ensure that a local DELETE after a local INSERT that is being re-deliverred does not re-introduce new tags). Is that what you're doing here?

I would expect that the merge of two entries would generate a new entry with tag being the the union of local and remote tag minus the local clear tags (in this case tag, would have to be a set --- I think this is a valid case: local insert with remote insert). 

ClearTags, I think can always be an union, because they just remove stuff that has been seen.

Somehow, it seems you're doing something different and I need to understand better.

[the remark this answers is not in the document]
Comment (balegas) [reply src/satellite/merge.ts:61]: I think this is the most critical part of this PR and I want to stop review here before continuing, so let me also make another comment: I think we need more unit tests to test more isolated bits of the code. Current tests are hard to follow for someone coming in. I have presented at least a couple of counter examples, if I am correct in some of my examples, we should have unit tests for showing the issue and use that as inspiration to make more tests.

File: src/satellite/process.ts [hunk 8]
```diff
@@ -489,26 +717,39 @@ export class SatelliteProcess implements Satellite {
           incomingChanges.optype
         )
 
-        Object.assign(incomingChanges, { changes, optype })
+        /*
+        console.log('merge_op_tag')
+        console.log(localChanges)
+        console.log(incomingChanges)
+         */
+        const clear_tags = mergeOpTags(localChanges, incomingChanges)
```
Comment (balegas) [inline src/satellite/process.ts:725]: Somehow I think you're doing the right thing, but I do not understand why this is called clear_tags and is making a lot of confusion to me (as of my previous long comment). Let's discuss this again.

**Review by balegas — CHANGES_REQUESTED** [review]
Minor comments and need to understand better your implementation of entries merging. I suppose this was a long time ago for me too :).

As a general comment, we need to have more unit tests for specific parts of the code, instead of testing everything top down

File: src/satellite/oplog.ts [hunk 9]
```diff
@@ -193,6 +227,7 @@ export const toTransactions = (
       relation: relations[`${entry.tablename}`],
       record,
       oldRecord,
+      tags: entry.optype == OPTYPES.delete ? [] : decodeTags(entry.clear_tags),
```
[the remark this answers is not in the document]
Comment (balegas) [reply src/satellite/oplog.ts:230]: After finishing the review, I imagine this might be right, but I just don't understand really the meaning of clear tags.

File: src/satellite/oplog.ts [hunk 10]
```diff
@@ -100,18 +124,21 @@ export const entryToChanges = (entry: OplogEntry): OplogEntryChanges => {
...
 export const operationsToTableChanges = (
-  operations: OplogEntry[]
+  operations: OplogEntry[],
+  genTag: (timestamp: Date) => Tag
 ): OplogTableChanges => {
   const initialValue: OplogTableChanges = {}
 
   return operations.reduce((acc, entry) => {
-    const entryChanges = entryToChanges(entry)
+    const entryChanges = entryToChanges(entry, () =>
```
[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/oplog.ts:136]: I'm not sure I follow what you mean

[the remark this answers is not in the document]
Comment (balegas) [reply src/satellite/oplog.ts:136]: I wonder if you could change the interface to not receive the lambda for generating the tag.

File: src/satellite/process.ts [hunk 11]
```diff
@@ -532,14 +773,18 @@ export class SatelliteProcess implements Satellite {
     ]
 
     if (isAck) {
+      /* Currenly we do not clean oplog table on acknowledgement due
```
[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/process.ts:776]: I left a comment to have an opportunity to discuss this further. 

With current version of GC we will keep oplog entries for longer time frames then we used to have, 
which will causes extra performance penalty when syncing transactions from Electric, and extra space
consumed on the device as we do not merge entries in the oplog, and they going to be GCed only on 
receiving same transaction from Electric and not just when Electric have confirmed that transaction has
been written. Also sending the whole transaction back to the source looks like an overkill in this situation,
 so it's also something for us to consider for optimization.

File: src/satellite/oplog.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/oplog.ts:136]: Lambda is used to capture the origin field, which is not present in the oplogEntry

File: src/satellite/client.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/client.ts:638]: The comment in proto file explicitly says about it:
```
   // Globally unique id of the source that transaction originated from. For
    // data coming from Satellite this field is ignored. For data coming from
    // Electric this field can be used to deduce if the incoming transaction
    // originated on this Satellite instance or not.
```
Let me know if further clarification on the protocol is desired

File: src/satellite/process.ts [hunk 12]
```diff
@@ -633,6 +879,23 @@ export class SatelliteProcess implements Satellite {
 
     return Promise.resolve(relations)
   }
+
+  private _generateTag(timestamp: Date): string {
+    const instanceId = this._authState!.clientId
+    return generateTag(instanceId, timestamp)
+  }
+
+  async _garbageCollectOplog(transaction: Transaction) {
```
[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/process.ts:889]: > but we could piggyback the ack with the transaction message.
I'm not sure what you mean here. Ack to the transaction and sending transaction back are decoupled processes and the underlying system doesn't have matching between lsn and transaction.

> Second, I think it looks unsafe to rely on timestamp,
As long as timestamp is strictly monotonically increasing for the target it should be safe. If it doesn't - then conflict resolution wouldn't work properly. We can store mapping `transaction -> lsn position` on Satellite if you think that's more appropriate.

File: src/satellite/merge.ts [hunk 13]
```diff
@@ -33,10 +40,37 @@ export const mergeChangesLastWriteWins = (
 }
 
 // Merge the type of two operations so that add wins.
-export const mergeOpTypesAddWins = (first: OpType, second: OpType): OpType => {
+export const mergeOpTypesAddWins = (
+  first: OpType | ChangesOpType,
```
[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/merge.ts:44]: I think merging opType and ChangesOpType requires a bit of extra work which is not really relevant for this PR.
They are used for different purposes - for oplog entries and for merging.

File: src/satellite/oplog.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (balegas) [reply src/satellite/oplog.ts:136]: I'm sorry to insist, but the param for getTag is a Date object from entry.timestamp in every reference to getTag that I could find.

[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/oplog.ts]: Now I see what you meant, fixed.

File: src/satellite/process.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (balegas) [reply src/satellite/process.ts]: >  Also sending the whole transaction back to the source looks like an overkill in this situation,
so it's also something for us to consider for optimization.
We could just send a gist of the transaction and not the transaction itself, but we need to verify the correctness of that. Right now, the transaction that we receive might not be exactly the one that we've sent (it will be the result of applying it on the server against concurrent outstanding operations). We would have to ensure that there are no corner cases in skipping the transaction... I am not very concerned with avoiding transmitting the transaction, because we have to send it to all other satellites anyways.

> With current version of GC we will keep oplog entries for longer time frames
We can eventually compact the log by merging entries that are safe to merge and other strategies. Compaction would be very efficient at least in the worst case for growing the oplog, which is when the client is disconnected.

I hope that the oplog doesn't grow so large when the user is online, otherwise it will not be a very 'real-time' system. In that sense I believe that the cost of the oplog should be much smaller compared to the cost of keeping application data on the device.

File: src/satellite/process.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (balegas) [reply src/satellite/process.ts:889]: >  Ack to the transaction and sending transaction back are decoupled processes and the underlying system doesn't have matching between lsn and transaction.

In the new approach, I think there could be no explicit ack message, because it is implied by receiving the message from the same origin. We would need to have a way to associate that with LSN, but this could be done by extending the transaction message to contain an optional ack field with the corresponding lsn position (which I don't know, but suppose is available on the server).

File: src/satellite/oplog.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/oplog.ts]: You are correct that we should not have a special case for delete here, I confused this with the scenario when data is coming from Electric, as there OPTYPES.delete means no tags. As for the entry.timestamp this tag is implicit based on the transaction info, so we optionally generate it when necessary

File: src/satellite/merge.ts [hunk 14]
```diff
@@ -33,10 +40,37 @@ export const mergeChangesLastWriteWins = (
...
+  second: OpType | ChangesOpType
+): ChangesOpType => {
   if (first === OPTYPES.delete && second === OPTYPES.delete) {
```
[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/merge.ts:47]: PR was implemented under assumption that we never reapply incoming transactions, that was also motivation for doing GC the way we do. If we need to reapply as you pointed out in another comment (which it seems we have to) then this logic is wrong.

[the remark this answers is not in the document]
Comment (define-null) [reply src/satellite/merge.ts:48]: Removed this logic, as it's no longer valid for tags approach.

File: src/satellite/process.ts [hunk 15]
```diff
@@ -713,17 +709,19 @@ export class SatelliteProcess implements Satellite {
...
-        )
+        let optype
 
         /*
         console.log('merge_op_tag')
         console.log(localChanges)
         console.log(incomingChanges)
          */
         const clearTags = mergeOpTags(localChanges, incomingChanges)
+        if (clearTags.length == 0) {
```
Comment (balegas) [inline src/satellite/process.ts:720]: I still find it confusing that this is called clearTags because what determines the optype is the presence of shadow tag and why in the `apply()` we also set tags with clearTags. 

We need to test this functions in isolation please. Could you make a simpler version of some of your tests that just generates the entries, calls apply, and checks what is the vicinity of `'INSERT wins over DELETE and restored deleted values'

File: src/satellite/process.ts [hunk 16]
```diff
@@ -301,38 +316,66 @@ export class SatelliteProcess implements Satellite {
...
+        console.log("step:")
+        console.log(oplogEntry)
+        console.log(shadowEntry)
+        */
+
+        // Clear should not contain the tag for this timestamp, so if
+        // the entry was previously in cache - it means, that we already
+        // read it within the same snapshot
+        if (cached) {
+          oplogEntry.clearTags = encodeTags(
```
Comment (balegas) [inline src/satellite/process.ts:350]: I think this is okay, but let's just discuss to make sure we understand:
- The case when we end up with the tag in the clear set and in the tag set is INSERT.INSERT
- Is it a problem to have the tag in clear and tag sets? No, if when applying an incoming operation we first remove the clear set and then add the tag.

Having said that, I don't disagree with this code.

File: src/satellite/process.ts [hunk 17]
```diff
@@ -25,7 +25,8 @@ import {
 } from '../util/types'
 import { Satellite } from './index'
 import { SatelliteConfig, SatelliteOpts } from './config'
-import { mergeChangesLastWriteWins, mergeOpTypesAddWins } from './merge'
+import { mergeChangesLastWriteWins, mergeOpTags } from './merge'
```
Comment (balegas) [inline src/satellite/process.ts:28]: In this function, couldn't we take advantage of the source to fix the issue with timestamp comparison?
(it seems it still has the issue when timestamps are equal)

File: src/satellite/merge.ts [hunk 18]
```diff
@@ -33,10 +41,28 @@ export const mergeChangesLastWriteWins = (
 }
 
 // Merge the type of two operations so that add wins.
-export const mergeOpTypesAddWins = (first: OpType, second: OpType): OpType => {
+export const mergeOpTypesAddWins = (
```
Comment (balegas) [inline src/satellite/merge.ts:44]: No longer used

File: src/satellite/oplog.ts [hunk 19]
```diff
@@ -131,6 +159,12 @@ export const operationsToTableChanges = (
       for (const [key, value] of Object.entries(entryChanges.changes)) {
         existing.changes[key] = value
       }
+      if (entryChanges.optype == 'DELETE') {
+        existing.tag = null
+      } else {
+        existing.tag = genTag(new Date(entry.timestamp))
+      }
+      existing.clearTags = entryChanges.clearTags
```
Comment (balegas) [inline src/satellite/oplog.ts:167]: If I understand correctly, when we compute a snapshot, we compute the clear set per entry and here we are taking the latest clear set from all entries.
I think we could have the following case:
t=[1], c=[]
t=[], c=[1]
t=[2], c=[]

And we would lose the clear of tag 1. Please provide a counter example if this is not a correct interpretation.

Comment (define-null) [reply src/satellite/oplog.ts]: fixed

File: src/satellite/process.ts (hunk 17 above)
Comment (define-null) [reply src/satellite/process.ts:28]: yes, that was the plan originally, fixed

File: test/support/satellite-helpers.ts [hunk 20]
```diff
@@ -60,6 +90,14 @@ export const generateOplogEntry = (
   if (optype === OPTYPES.update || optype === OPTYPES.delete) {
     oldRow = generateFrom(schema, oldValues)
   }
+  let tags = clearTags
+  // I'd prefer much more to always have these as parameter
```
Comment (define-null) [inline test/support/satellite-helpers.ts:94]:
```suggestion
```

File: src/satellite/process.ts [hunk 21]
```diff
@@ -301,38 +317,66 @@ export class SatelliteProcess implements Satellite {
...
+        )
 
-    const promises: Promise<void | SatelliteError>[] = []
+        /*
+        console.log("shadowentries:")
+        console.log(shadowEntries)
+        console.log("step:")
+        console.log(oplogEntry)
+        console.log(shadowEntry)
+        */
```
Comment (define-null) [inline src/satellite/process.ts:345]:
```suggestion
```

File: src/satellite/process.ts [hunk 22]
```diff
@@ -301,38 +317,66 @@ export class SatelliteProcess implements Satellite {
...
+        } else {
+          const newTag = this._generateTag(timestamp)
+          shadowEntry.tags = encodeTags([newTag])
+        }
+
+        /*
+        console.log("after:")
+        console.log(oplogEntry)
+        console.log(shadowEntry)
+        */
```
Comment (define-null) [inline src/satellite/process.ts:371]:
```suggestion
```

File: src/satellite/process.ts [hunk 23]
```diff
@@ -394,16 +478,32 @@ export class SatelliteProcess implements Satellite {
...
+      incoming
+    )
+
+    /*
+    console.log(local)
+    console.log(incoming)
+
+    console.log('merged:')
+    console.log(merged)
+     */
```
Comment (define-null) [inline src/satellite/process.ts:506]:
```suggestion
```

File: src/satellite/process.ts [hunk 24]
```diff
@@ -418,13 +518,24 @@ export class SatelliteProcess implements Satellite {
...
       for (const entryChanges of Object.values(mapping)) {
+        const shadowEntry: ShadowEntry = {
+          namespace: entryChanges.namespace,
+          tablename: entryChanges.tablename,
+          primaryKey: getShadowPrimaryKey(entryChanges),
+          tags: encodeTags(entryChanges.tags),
+        }
+        //console.log('shadow:')
+        //console.log(shadowEntry)
+
```
Comment (define-null) [inline src/satellite/process.ts:529]:
```suggestion
```

File: src/satellite/process.ts [hunk 25]
```diff
@@ -474,41 +709,70 @@ export class SatelliteProcess implements Satellite {
...
-          localChanges.optype,
-          incomingChanges.optype
-        )
+        let optype
+
+        /*
+        console.log('merge_op_tag')
+        console.log(localChanges)
+        console.log(incomingChanges)
+         */
```
Comment (define-null) [inline src/satellite/process.ts:730]:
```suggestion
```

File: test/satellite/process.tags.test.ts [hunk 26]
```diff
@@ -0,0 +1,790 @@
...
+  t.is(entries[1].clearTags, genEncodedTags(clientId, [txDate1]))
+  t.is(entries[2].clearTags, genEncodedTags(clientId, [txDate2]))
+  t.is(entries[3].clearTags, genEncodedTags(clientId, [txDate3]))
+
+  t.not(txDate1, txDate2)
+  t.not(txDate2, txDate3)
+  t.not(txDate3, txDate4)
+})
+
+// In this test we take three snapshots for three separate transactions.
```
Comment (balegas) [inline test/satellite/process.tags.test.ts:149]: Please remove the long explanation (which was meant to you) and adapt the test name accordingly with other tests you created.

File: test/satellite/process.tags.test.ts [hunk 27]
```diff
@@ -0,0 +1,790 @@
...
+  t.like(
+    shadow[0],
+    {
+      tags: genEncodedTags(clientId, [txDate3]),
+    },
+    'error: tag1 was reintroduced after merging acked operation'
+  )
+})
+
+// Problem of previous test does not occur with just two operations: INSERT, DELETE
```
Comment (balegas) [inline test/satellite/process.tags.test.ts:260]: This and the following tests are not necessary since the intent was just to show the issue would not occur in scenarios with just two ops.

File: src/satellite/oplog.ts [hunk 28]
```diff
@@ -121,19 +140,45 @@ export const entryToChanges = (
...
+  return result
+}
+
 // Convert a list of `OplogEntry`s into a nested `OplogTableChanges` map of
 // `{tableName: {primaryKey: entryChanges}}` where the entryChanges has the
 // most recent `optype` and column `value`` from all of the operations.
 // Multiple OplogEntries that point to the same row will be merged to a
 // single OpLogEntryChanges object.
-export const operationsToTableChanges = (
+export const localOperationsToChanges = (
```
Comment (balegas) [inline src/satellite/oplog.ts:174]: Changes/TableChanges should be consistent with type name

File: src/satellite/merge.ts [hunk 29]
```diff
@@ -1,17 +1,17 @@
...
   Tag,
   OplogEntryChanges,
+  ShadowEntryChanges,
 } from './oplog'
 import { difference, union } from '../util/sets'
 
 // Merge two sets of changes, using the timestamp to abitrate conflicts
 // so that the last write wins.
 export const mergeChangesLastWriteWins = (
+  first_origin: string,
```
Comment (balegas) [inline src/satellite/merge.ts:12]: watch for naming convention

**Review by balegas — APPROVED** [review]
I think it looks goods.

**Comment by define-null** [comment]
Rebased on latest main.

File: test/satellite/process.tags.test.ts (hunk 27 above)
Comment (define-null) [reply test/satellite/process.tags.test.ts]: removed

File: test/satellite/process.tags.test.ts (hunk 26 above)
Comment (define-null) [reply test/satellite/process.tags.test.ts]: removed

**Review by balegas — APPROVED** [review]
Looks good to me. Great work.
Please close VAX-391 too.
